### PR TITLE
[DO NOT MERGE] Consistent handling of semicolons in macro expansions

### PR DESCRIPTION
**Pull request by Aaron1011**
Implements 'Proposal 2' from https://github.com/rust-lang/rust/issues/61733#issuecomment-509626449

This PR treats semicolon-less macro invocations like statements. This allows the following code to compile:

```rust
fn main() {
    macro_rules! a {
        ($e:expr) => { $e; }
    }
    a!(true)
}
```

To avoid making macro behavior more inconsistent, macros expanded in expression position no longer ignore trailing semicolons in the expansion. This broke several things in the standard library - in fact, we can't even get rustfmt to pass until https://github.com/rust-lang/rustfmt/pull/4507 is merged

Given all of the issues this has expoesd with the standard library (and many test cases), I expect this to cause a significant amount of ecosystem breakage. I'm opening this for a Crater run - we'll almost certainly need to do a gradual rollout if this is accepted. Fortunately, it's easy to convert the `macro_rules!` error into a warning.

### CODE REVIEW

File: library/core/src/str/validations.rs [hunk 1]
```diff
@@ -125,7 +125,7 @@ pub(super) fn run_utf8_validation(v: &[u8]) -> Result<(), Utf8Error> {
         let old_offset = index;
         macro_rules! err {
             ($error_len: expr) => {
-                return Err(Utf8Error { valid_up_to: old_offset, error_len: $error_len });
+                return Err(Utf8Error { valid_up_to: old_offset, error_len: $error_len })
```
Comment (o752d) [inline library/core/src/str/validations.rs:128]:
```suggestion
                Err(Utf8Error { valid_up_to: old_offset, error_len: $error_len })
```
Is `return` required here?

Comment (Aaron1011) [reply library/core/src/str/validations.rs:128]: Yes - for example, https://github.com/Aaron1011/rust/blob/003157547b12ef2b81db5aaa223070614b3a4327/library/core/src/str/validations.rs#L136-L138 uses this to bail out early.

**Comment by Aaron1011** [comment]
The build doesn't even complete because of `anyhow`:

https://github.com/dtolnay/anyhow/blob/c25be95f1a24f7497d2b4530ffcb4f90d3871975/src/macros.rs#L54-L64

Opened https://github.com/dtolnay/anyhow/pull/120. The fact that three different crates (`libstd`, `stdarch`, and `anyhow`) have all been broken by this change doesn't seem to bode well for the Crater results.

**Comment by petrochenkov** [comment]
Closing due to inactivity.

**Pull request closed without merging**